akka-stream/src/main/scala/akka/stream/stage/GraphStage.scala

```diff
@@ -1317,43 +1318,33 @@ abstract class GraphStageLogic private[stream] (val inCount: Int, val outCount:
 
   // Internal hooks to avoid reliance on user calling super in postStop
   /** INTERNAL API */
+  @silent("JavaConverters in package collection is deprecated")
   protected[stream] def afterPostStop(): Unit = {
     if (_stageActor ne null) {
       _stageActor.stop()
       _stageActor = null
     }
+
+    import scala.collection.JavaConverters._
     // make sure any invokeWithFeedback after this fails fast
     // and fail current outstanding invokeWithFeedback promises
-    val inProgress = asyncCallbacksInProgress.getAndSet(null)
+    val inProgress =
+      asyncCallbacksInProgress
+        .getAndSet(null) // remove reference to signify that we are in the process of shutting down
+        .keys()
+        .asScala
     if (inProgress.nonEmpty) {
       val exception = streamDetachedException
       inProgress.foreach(_.tryFailure(exception))
     }
   }
 
-  private[this] var asyncCleanupCounter = 0L
-
   /** Called from interpreter thread by GraphInterpreter.runAsyncInput */
-  private[stream] def onFeedbackDispatched(): Unit = {
-    asyncCleanupCounter += 1
-
-    // 256 seemed to be a sweet spot in SendQueueBenchmark.queue benchmarks
-    // It means that at most 255 completed promises are retained per logic that
-    // uses invokeWithFeedback callbacks.
-    //
-    // TODO: add periodical cleanup to get rid of those 255 promises as well
-    if (asyncCleanupCounter % 256 == 0) {
-      @tailrec def cleanup(): Unit = {
-        val previous = asyncCallbacksInProgress.get()
-        if (previous != null) {
-          val updated = previous.filterNot(_.isCompleted)
-          if (!asyncCallbacksInProgress.compareAndSet(previous, updated)) cleanup()
-        }
-      }
-
-      cleanup()
+  private[stream] def onFeedbackDispatched(p: Promise[Done]): Unit =
+    asyncCallbacksInProgress.get() match {
+      case null => // already finished, nothing to do here
+      case x    => x.remove(p)
```

Comment:
better than the left, but how about wrap the ConcurrentHashMap as ConcurrentHashSet?

Reply:
WDYM? Is there a `ConcurrentHashSet` somewhere?

Reply:
I could use `ConcurrentHashMap.newKeySet`, trying that...

Reply:
I mean `left` the old implementation:)